packages/gatsby-plugin-sharp/src/gatsby-node.js

```diff
@@ -1,10 +1,93 @@
-const { setBoundActionCreators, setPluginOptions } = require(`./index`)
+const {
+  setBoundActionCreators,
+  setPluginOptions,
+  queue,
+  reportError,
+} = require(`./index`)
+const processFile = require(`./processFile`)
+const { scheduleJob } = require(`./scheduler`)
 
-exports.onPreInit = ({ actions }, pluginOptions) => {
+const getQueueFromCache = async cache => {
+  const rawQueue = await cache.get(`queue`)
+
+  if (!rawQueue) {
+    return new Map()
+  }
+
+  return new Map(rawQueue)
+}
+
+const saveQueueToCache = async cache => {
+  const cachedQueue = await getQueueFromCache(cache)
+
+  // merge both queues
+  for (const [key, job] of cachedQueue) {
+    queue.set(key, job)
+  }
+
+  // JSON.stringify doesn't work on an Map so we need to convert it to an array
+  return cache.set(`queue`, Array.from(queue))
+}
+
+exports.onPreInit = async ({ actions }, pluginOptions) => {
   setBoundActionCreators(actions)
   setPluginOptions(pluginOptions)
 }
 
+/**
+ * save queue to the cache
+ */
+exports.onPostBootstrap = async ({ cache, store }) => saveQueueToCache(cache)
+
+/**
+ * Execute all unprocessed images on gatsby build
+ */
+let promises = []
+exports.onPreBuild = async ({ cache, boundActionCreators }, pluginOptions) => {
+  const cachedQueue = await getQueueFromCache(cache)
+
+  for (const [, job] of cachedQueue) {
+    promises.push(scheduleJob(job, boundActionCreators, pluginOptions))
+  }
+}
+
+/**
+ * wait for all images to be processed
+ */
+exports.onPostBuild = ({ cache, reporter }) =>
+  Promise.all(promises)
+    .then(() => cache.set(`queue`, []))
+    .catch(({ err, message }) => {
+      reportError(message || err.message, err, reporter)
+    })
+
+/**
+ * Build images on the fly when they are requested by the browser
+ */
+exports.onCreateDevServer = async ({ app, cache, compiler }, pluginOptions) => {
+  compiler.hooks.done.tap(`gatsby-plugin-sharp`, () => {
```

Comment:
In need to hook in on the compilation lifcycle to know whenever a new build is done.

Reply:
peer-review with @m-allanson 

> lets go for emitter.on('QUERY_QUEUE_DRAINED') instead of exposing compiler